Bootstrap most of Spack dependencies using environments

alalazo opened the pull request:
The main objective of this PR is to rework the bootstrapping procedure to use Spack environments as much as possible. 

While doing so, the `spack.bootstrap` module has been reorganized into a Python package. For reviewers, the new code is under `spack.bootstrap.environment` (most of the other changes are just adding docs or linting).

In this PR a distinction is made among "core" Spack dependencies (clingo, GnuPG, patchelf) and other dependencies. For a number of reasons, explained in the `spack.bootstrap.core` module docstring, "core" dependencies are bootstrapped with the current ad-hoc method. All the other dependencies are instead bootstrapped using a Spack environment that lives in a directory specific to the interpreter and the architecture being used. An example of the directory structure is:
```console
$ tree $(spack bootstrap root) -L 3
/home/culpo/.spack/bootstrap
├── config
│   └── linux
│       ├── bootstrap.yaml
│       ├── compilers.yaml
│       └── config.yaml
├── environments
│   └── python3.8-x86_64-57f95
│       ├── Makefile
│       ├── spack.lock
│       ├── spack.yaml
│       └── view -> /home/culpo/.spack/bootstrap/environments/python3.8-x86_64-57f95/._view/dgsyi3525yl3vzeakomxqrekchzd7l6s
└── store
    ├── bin
    │   └── sbang
    ├── linux-centos7-x86_64
    │   └── gcc-10.2.1
    └── linux-ubuntu20.04-x86_64
        └── gcc-9.4.0
```
The bootstrapping output has been reduced and now looks like:
![Screenshot from 2022-11-21 07-00-59](https://user-images.githubusercontent.com/4199709/202976861-049432ed-a5d5-47d3-b2d7-fadef1ca3684.png)

Modifications:
- [x] Bootstrap most of Spack dependencies using a Spack environment
- [x] Add `pytest` to the list of bootstrapped dependencies
- [x] Reduced verbosity of bootstrapping output
- [x] Added a `--dev` argument to `spack bootstrap now`
- [x] Reorganized `spack.bootstrap` into a package and linted it

alalazo committed:
Move bootstrap.py inside a package

alalazo committed:
Extract functions related to swapping configuration in their own module

alalazo committed:
Extract functions related to showing the bootstrap status in their own module

alalazo committed:
Have a submodule for the ad-hoc bootstrapping procedure from binaries

alalazo committed:
Linted the spack.bootstrap package with pylint defaults

alalazo committed:
Install dev-dependencies with a Spack environment

alalazo committed:
Add pytest back as a dependency

alalazo committed:
spack.bootstrap: reduce the number of exported names

alalazo committed:
Collect common methods in a class

alalazo committed:
Decrease verbosity of the output

alalazo committed:
Another run of linting and polishing

alalazo commented:
A few notes in addition to the description:
1. This PR makes it possible to set the bootstrap store and the user store in the same place. I still think it's a very bad idea to mix installations needed by Spack for its own working with installations needed by the user, but if people want to experiment with that it is now possible (@psakievich might be interested given the discussions during community meetings)
2. After this PR is in, we can delete `pytest` from the Python packages we vendor (see #33951)

alalazo committed:
Conform to prescribed number of jobs, add debug output back

alalazo committed:
Fix circular import on macOS

alalazo committed:
Python.command: try platform-python on rhel8 as a last resort

alalazo committed:
Fix reference before assignment on Windows

alalazo committed:
Avoid requiring Python headers

This is a hack to install from wheels two
Python packages. We might do better in the future.

psakievich commented:
@alalazo I agree with you that there are a lot of dangers in mixing the stores. I would primarily like to reuse the bootstrapped dependencies from `spack style` and soon to be `spack unit-test`.  I'm not totally sure of what else is involved beyond those and the "core" dependencies you've outlined in the PR description. I don't have the time to really dig into a PR this big right now so I'll ask a couple of questions instead.

1) Are the "core" dependencies (clingo, patchelf, GnuPG) included in the store mixing?  I think it would be best to keep those in a separate store/install_tree, and keep them off limits.  From reading the comments and doc string I believe that is correct?
2) How is the "store mixing" done?  Will `spack uninstall --all` blow away the bootstrap store, or will it be treated like an upstream with some level of additional protection?  I guess if it's just some python packages that get blown away then I can live with that.  If it is something more fundamental, then that is less ideal.

alalazo commented:
@psakievich 

> I'm not totally sure of what else is involved beyond those and the "core" dependencies you've outlined in the PR description.

https://github.com/spack/spack/blob/d5142d1cb651705113c02b8b48b03ddb27b950e7/lib/spack/spack/bootstrap/core.py#L5-L23

> Are the "core" dependencies (clingo, patchelf, GnuPG) included in the store mixing? [ ... ]

To be clear:
1. There is no mixing by default, things remain exactly as on `develop` with that respect. I just removed a check that was preventing people from setting the "bootstrapping store" and `config:install_tree:root` to the same place.
2. The mixing is done as you like, if you want to experiment with that. You can use the bootstrap store as an upstream, or you can set both stores to the same place.[^1]

[^1]: I wouldn't recommend any of the two, but they are possible

haampie commented on lib/spack/spack/bootstrap/environment.py:
```diff
@@ -0,0 +1,191 @@
+            "-C",
+            str(self.environment_root()),
+            "-j",
+            str(spack.build_environment.determine_number_of_jobs(parallel=True)),
```
One issue here is that it will regenerate the view for every package being installed

alalazo replied:
This is in any case much faster than a serial installation, so I guess we can optimize later to generate a depfile that can install without generating the view?[^1]

[^1]: It's also ugly that we have to spawn another process to generate the Makefile, but given the difference in performance I think it's preferable to `env.install_all()`.

alalazo committed:
Canonicalize the root path

alalazo committed:
Remove the external Python from the environment.py

For some reason to be debugged later this triggers an in-memory bug
with the first installation with env.install_all().

To be investigated why the bug is triggered.

alalazo committed:
Add back missing dependencies after #33898

trws commented on lib/spack/spack/bootstrap/config.py:
```diff
@@ -0,0 +1,168 @@
+def _ensure_bootstrap_configuration():
+    bootstrap_store_path = store_path()
+    user_configuration = _read_and_sanitize_configuration()
+    with spack.environment.no_active_environment():
+        with spack.platforms.prevent_cray_detection():
+            with spack.platforms.use_platform(spack.platforms.real_host()):
+                with spack.repo.use_repositories(spack.paths.packages_path):
+                    with spack.store.use_store(bootstrap_store_path):
```
Thoughts on maybe collapsing this to a single `with` statement with multiple values? The semantics are the same, but avoiding the extra four levels of nesting might be nice.

trws commented on var/spack/repos/builtin/packages/py-black/package.py:
```diff
@@ -52,6 +52,9 @@ class PyBlack(PythonPackage):
     depends_on("py-tokenize-rt@3.2:", when="+jupyter", type=("build", "run"))
 
+    # Needed because this package is used to bootstrap Spack (Spack supports Python 3.6+)
+    depends_on("py-dataclasses@0.6:", when="^python@:3.6", type=("build", "run"))
```
Are we supporting anything this would match for anymore? I may be misreading but this only applies below 3.6 right?

trws replied:
Nevermind, I misread it, sorry.

trws commented:
All my testing shows this working well, mac, linux, arm, x86, I have yet to find a problem with bootstrapping.  The one thing that struck me as a bit odd though, I'm getting output describing what's going to be built *after* it's built rather than before but I'm not seeing obviously why that is:

```
==> clingo-bootstrap: Successfully installed clingo-bootstrap-spack-ls7pgf6b6ytnlgtgim4cpdsfvm4hrmqo
  Stage: 0.33s.  Cmake: 2.03s.  Build: 28.68s.  Install: 0.55s.  Total: 31.73s
[+] /Users/scogland1/.spack/bootstrap/store/darwin-monterey-aarch64/apple-clang-14.0.0/clingo-bootstrap-spack-ls7pgf6b6ytnlgtgim4cpdsfvm4hrmqo
==> [BOOTSTRAPPING] Spack has missing dependencies, creating a bootstrapping environment
==> [BOOTSTRAPPING] Installing dependencies (py-isort@4.3.5:, py-mypy@0.900:, py-black, py-flake8, py-pytest)
```

alalazo committed:
Remove nesting of context managers

alalazo committed:
Make flake8 checks pass

alalazo committed:
Fix unit test that leaks environment modifications

alalazo commented:
Took me a while to find, but the unit tests failing [here](https://github.com/spack/spack/actions/runs/3581628451/jobs/6025916090) are because some environment modification leak, fixed in b8f0964, of tests running before them. The unit tests have been modified in #30882

> The one thing that struck me as a bit odd though, I'm getting output describing what's going to be built after it's built rather than before but I'm not seeing obviously why that is: [ ... ]

That's because `clingo`, ` patchelf` and `GnuPG` are treated differently. We first bootstrap them with an ad-hoc procedure, then use the environment for everything else (currently py-isort@4.3.5:, py-mypy@0.900:, py-black, py-flake8, py-pytest) - so what you see is the bootstrapping of "core" software before the environment.

trws commented:
Got it, doing some more testing, but I'm reminded of a related bootstrap question.  Do we need to keep using a development version of clingo? As it is, we're forcing an increase in build dependencies to do it, is there no release new enough to get rid of the re2c and bison dependencies?  Especially with an external cmake and python available, the size of the build set is substantially larger because of it:

```
# spack version
clingo-bootstrap@spack%apple-clang@14.0.0~docs~ipo+python~static_libstdcpp build_system=cmake build_type=Release arch=darwin-monterey-m1
    ^bison@3.8.2%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
        ^diffutils@3.8%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
            ^libiconv@1.16%apple-clang@14.0.0 build_system=autotools libs=shared,static arch=darwin-monterey-m1
        ^gnuconfig@2022-09-17%apple-clang@14.0.0 build_system=generic arch=darwin-monterey-m1
        ^m4@1.4.19%apple-clang@14.0.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=darwin-monterey-m1
            ^libsigsegv@2.13%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
        ^perl@5.36.0%apple-clang@14.0.0+cpanm+shared+threads build_system=generic arch=darwin-monterey-m1
            ^berkeley-db@18.1.40%apple-clang@14.0.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=darwin-monterey-m1
            ^bzip2@1.0.8%apple-clang@14.0.0~debug~pic+shared build_system=generic arch=darwin-monterey-m1
            ^gdbm@1.23%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
                ^readline@8.1.2%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
                    ^ncurses@6.3%apple-clang@14.0.0~symlinks+termlib abi=none build_system=autotools arch=darwin-monterey-m1
                        ^pkgconf@1.8.0%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
            ^zlib@1.2.13%apple-clang@14.0.0+optimize+pic+shared build_system=makefile arch=darwin-monterey-m1
    ^cmake@3.24.2%apple-clang@14.0.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=darwin-monterey-m1
    ^python@3.9.6%apple-clang@14.0.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma+nis~optimizations+pic+pyexpat~pythoncmd+readline+shared+sqlite3+ssl+tix+tkinter+uuid+zlib build_system=generic patches=0d98e93,f2fd060 arch=darwin-monterey-m1
    ^re2c@2.2%apple-clang@14.0.0 build_system=generic arch=darwin-monterey-m1

# release version
clingo-bootstrap@5.5.2%apple-clang@14.0.0~docs~ipo+python~static_libstdcpp build_system=cmake build_type=Release arch=darwin-monterey-m1
    ^cmake@3.24.2%apple-clang@14.0.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=darwin-monterey-m1
    ^py-cffi@1.15.0%apple-clang@14.0.0 build_system=python_pip arch=darwin-monterey-m1
        ^libffi@3.4.2%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
            ^gnuconfig@2022-09-17%apple-clang@14.0.0 build_system=generic arch=darwin-monterey-m1
        ^pkgconf@1.8.0%apple-clang@14.0.0 build_system=autotools arch=darwin-monterey-m1
        ^py-pip@22.2.2%apple-clang@14.0.0 build_system=generic arch=darwin-monterey-m1
        ^py-pycparser@2.21%apple-clang@14.0.0 build_system=python_pip arch=darwin-monterey-m1
        ^py-setuptools@65.5.0%apple-clang@14.0.0 build_system=generic arch=darwin-monterey-m1
        ^py-wheel@0.37.1%apple-clang@14.0.0 build_system=generic arch=darwin-monterey-m1
    ^python@3.9.6%apple-clang@14.0.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma+nis~optimizations+pic+pyexpat~pythoncmd+readline+shared+sqlite3+ssl+tix+tkinter+uuid+zlib build_system=generic patches=0d98e93,f2fd060 arch=darwin-monterey-m1

alalazo commented:
> Do we need to keep using a development version of clingo?

We are using our binaries most of the time, which are installing a single node. So the explosion in dependencies is there only when building from sources (this didn't change in this PR). That said, that particular version was a requirement up until v0.19.0 (last clingo commit supporting Python 2.7). We can look into updating clingo for v0.20.0.

If you don't use binaries on `m1`, do you know if you fall in one of these cases https://github.com/spack/spack/issues/33987 ?

trws commented:
I'm intentionally bootstrapping from source to test everything.  We can certainly look into updating clingo separately, just avoiding that darned perl build would be a bonus.  Bit of a surprise on an ubuntu container too actually, apparently we don't bootstrap curl but try to use it?  We also don't check for ca-certificates existing, so we get curl missing and try `spack -k`, but only when it hits patchelf (gnuconfig installs without either). 

```
==> Installing patchelf-0.16.1-cljjfq4ynfu2txmln6mbcn4kxv2jfvj7
==> No binary for patchelf-0.16.1-cljjfq4ynfu2txmln6mbcn4kxv2jfvj7 found: installing from source
==> Error: spack requires 'curl'. Make sure it is in your path.
==> Error: spack requires 'curl'. Make sure it is in your path.
==> Error: spack requires 'curl'. Make sure it is in your path.
==> Error: spack requires 'curl'. Make sure it is in your path.
==> Warning: Spack was unable to fetch url list due to a certificate verification problem. You can try running spack -k, which will not check SSL certificates. Use this at your own risk.
```

alalazo commented:
> Bit of a surprise on an ubuntu container too actually, apparently we don't bootstrap curl but try to use it? 

`curl` is not managed by the bootstrapping mechanism currently. I thought we relied on Python `urllib` by default, but something might have changed in #29026. Need to check that PR more closely.

trws commented:
Based on what I'm seeing it definitely tries urllib, but as a fallback from curl, and it doesn't try to grab ca-certificates if they're unavailable.  Normally we can reasonably expect those, but it might be nice to have a way to grab them if they're missing.

alalazo commented:
> Based on what I'm seeing it definitely tries urllib, but as a fallback from curl, and it doesn't try to grab ca-certificates if they're unavailable. Normally we can reasonably expect those, but it might be nice to have a way to grab them if they're missing.

I think it's a bug introduced in #29026. Pinged you in a discussion there too :slightly_smiling_face:

tldahlgren commented:
> > Based on what I'm seeing it definitely tries urllib, but as a fallback from curl, and it doesn't try to grab ca-certificates if they're unavailable. Normally we can reasonably expect those, but it might be nice to have a way to grab them if they're missing.
> 
> I think it's a bug introduced in #29026. Pinged you in a discussion there too 🙂

It is.  If no-one else is already fixing it I will.

alalazo commented:
@spackbot run pipeline

alalazo committed:
Fix import issue after rebase

alalazo commented:
@trws Is there any further request for changes?

alalazo commented:
> Would you be willing to take a stab at adding locking to the bootstrap.core end, perhaps shortly after this merges?

Sure!

alalazo commented:
@spackbot run pipeline

alalazo commented:
Merging, since the spec only that is reported broken is also broken in develop, see https://gitlab.spack.io/spack/spack/-/jobs/4817633

alalazo commented:
I'll work on fixing that spec next. For the record the pipeline finished here https://gitlab.spack.io/spack/spack/-/pipelines/233092 and so far never reported back to the PR status:
![Screenshot from 2022-12-06 11-51-05](https://user-images.githubusercontent.com/4199709/205891820-7d82fd31-afd0-4e5e-9381-09a779eb710b.png)

tgamblin commented:
I am not 100% sure it's this PR, but I am seeing a lot of random bootstrap failures in GHA and I think it may be due to this PR.
